.github/workflows/ci.yml

```diff
@@ -0,0 +1,148 @@
+name: CI
+
+on:
+  push:
+    branches: [ master ]
+    tags: [ "**" ]
+  pull_request:
+    branches: [ "**" ]
+
+defaults:
+  run:
+    shell: bash
+
+jobs:
+  build:
+    name: ${{ matrix.task.name}} - ${{ matrix.os.name }} ${{ matrix.python.name }}
+    runs-on: ${{ matrix.os.runs-on }}
+    strategy:
+      fail-fast: false
+      matrix:
+        os:
+          - name: Linux
+            runs-on: ubuntu-latest
+        python:
+          - name: CPython 3.8
+            tox: py38
+            action: 3.8
+        task:
+          - name: Build
+            tox: build
+
+    steps:
+    - uses: actions/checkout@v2
+
+    - name: Set up ${{ matrix.python.name }}
+      uses: actions/setup-python@v2
+      with:
+        python-version: ${{ matrix.python.action }}
+
+    - name: Install dependencies
+      run: |
+        python -m pip install --upgrade pip setuptools wheel
+        python -m pip install build check-manifest twine
+
+    - uses: twisted/python-info-action@v1
+
+    - name: Build
+      run: |
+        check-manifest --verbose .
+
+        python -m build --sdist --outdir dist/ .
+
+        mkdir empty/
+        cd empty
+
+        tar -xvf ../dist/*
+        cd *
+
+        # build the wheel from the sdist
+        python -m build --wheel --outdir ../../dist/ .
+        cd ../../
+
+        twine check dist/*
+
+    - name: Publish
+      uses: actions/upload-artifact@v2
+      with:
+        name: dist
+        path: dist/
+
+  test:
+    name: ${{ matrix.task.name}} - ${{ matrix.os.name }} ${{ matrix.python.name }}
+    runs-on: ${{ matrix.os.runs-on }}
+    needs:
+      - build
+    strategy:
+      fail-fast: false
+      matrix:
+        os:
+          - name: Linux
+            runs-on: ubuntu-latest
+        python:
+          - name: CPython 3.5
+            tox: py35
+            action: 3.5
+          - name: CPython 3.6
+            tox: py36
+            action: 3.6
+          - name: CPython 3.7
+            tox: py37
+            action: 3.7
+          - name: CPython 3.8
+            tox: py38
+            action: 3.8
+          - name: CPython 3.9
+            tox: py39
+            action: 3.9
+          - name: PyPy 3.6
+            tox: pypy36
+            action: pypy-3.6
+          - name: PyPy 3.7
+            tox: pypy37
+            action: pypy-3.7
+        task:
+          - name: Test
+            tox: tests
+
+    steps:
+    - uses: actions/checkout@v2
+
+    - name: Download package files
```

Comment:
Please correct me if I'm wrong -- I'm not familiar with testing Python packages this way.

Wouldn't it be simpler to just install dependencies and run the pre-commit script:
```
- name: Install dependencies
  run: pip3 install .[dev]

- name: Run checks
  run: python precommit.py
```
? This simplification would also allow you to dispense of `MANIFEST.in`?

Reply:
Let's see how to fix this in a separate pull request.

Reply:
Just installing from the source doesn't check to see if the built sdist and wheel work.  Usually it doesn't matter.  Sometimes you get into situations where you end up missing files in the sdist.  The purpose of this overall structure is to 1) create the artifacts you are going to publish (sdist and wheel) then 2) test that single wheel file in all places it might be used (that you cover in CI) and 3) publish those same exact artifacts that were built in the first step.  Then the thing to be published is the thing that gets tested.  Rather than doing several different installations from several different source checkouts and testing that followed by creating an sdist and wheel to upload that never get tested at all.

But, I'm going to unresolve this conversation and take a look at the precommit before moving on.

Reply:
I am used to tox being the central location for activity descriptions.  It looks like you are using precommit for that role.  If using `precommit.py`, I think it may make sense to add some more configurability (arguments) for it to be used here.  I'm used to linting activities (isort, mypy, yapf, etc) being run on the vcs copy and tests with coverage being run on the installed package.

`MANIFEST.in` is just about describing what should and shouldn't be in your sdist.  I almost certainly don't have the `MANIFEST.in` correct, but I don't think precommit is an alternative.

I'm going to phone a friend on this one since I don't have experience with precommit and what sort of setup works out well for all the use cases.  @graingert, if you have a couple minutes to help get me up to speed...  thanks.

Reply:
Thanks for the clarification re tests! I haven't spent too much thought on it -- it actually does make sense to include the tests so that users can test for the "dependency hell" and that everything works with *their* virtual environment which might have certain dependencies with different versions already installed.

We used `precommit.py` just because it was simple and convenient. If you are familiar with `tox`, please feel free to move the code into tox. In some packages I wanted to add more complex logic into precommits, so tox falls short in that regard. For this particular package, though, tox should be just fine.

Reply:
Should I merge the pull request as it is or you'd like to integrate first the pre-commit (tox or otherwise)?

Reply:
I guess go ahead and merge it.  As is, it is better than the present state of not having CI.  Maybe some of the other bits will be able to be tackled in focused PRs.  I'm certainly game for adding more tox, but I don't want to remove the existing precommit functionality that I presume you like.  I need to figure out how they play together without too much repetition either of code or of env management.

Reply:
Thanks, merged! You can just split up `precommit.py` into `tox.ini` as you see fit.